src/segment-spinner/SegmentState.ts

```diff
@@ -0,0 +1,164 @@
+/**
+ * All credit goes to [React Spectrum](https://github.com/adobe/react-spectrum)
+ * We improved the Calendar from Stately [useCalendarState](https://github.com/adobe/react-spectrum/tree/main/packages/%40react-stately/calendar)
+ * to work with Reakit System
+ */
+
+import { useCompositeState } from "reakit";
+import { useMemo, useState } from "react";
+import { useDateFormatter } from "@react-aria/i18n";
+import { useControlledState } from "@react-stately/utils";
+
+import { add, setSegment, convertValue, getSegmentLimits } from "./__utils";
+
+export interface IDateSegment {
+  type: Intl.DateTimeFormatPartTypes;
+  text: string;
+  value?: number;
+  minValue?: number;
+  maxValue?: number;
+  isPlaceholder: boolean;
+}
+
+const EDITABLE_SEGMENTS = {
+  year: true,
+  month: true,
+  day: true,
+  hour: true,
+  minute: true,
+  second: true,
+  dayPeriod: true,
+};
+
+const PAGE_STEP = {
+  year: 5,
+  month: 2,
+  day: 7,
+  hour: 2,
+  minute: 15,
+  second: 15,
+};
+
+// Node seems to convert everything to lowercase...
+const TYPE_MAPPING = {
+  dayperiod: "dayPeriod",
+};
+
+export interface SegmentStateProps {
+  value?: Date;
+  defaultValue?: Date;
+  formatOptions?:
+    | (Intl.DateTimeFormatOptions & { timeStyle?: string; dateStyle?: string })
+    | undefined;
```

Comment:
```suggestion
formatOptions?: Intl.DateTimeFormatOptions & {
    timeStyle?: string;
    dateStyle?: string;
  };
```

Reply:
I believe `undefined` is not needed because `?:` of optional type

Reply:
Weird thing i did not noticed this